src/option_learning.py

```diff
@@ -212,3 +219,205 @@ def update_segment_from_option_spec(self, segment: Segment,
+        object.__setattr__(self, "_operator", operator)
+        object.__setattr__(self, "_regressor", regressor)
+        object.__setattr__(self, "_changing_parameter_idxs",
+                           changing_parameter_idxs)
+        super().__init__(name,
+                         types,
+                         params_space,
+                         _policy=self._regressor_based_policy,
+                         _initiable=self._precondition_based_initiable,
+                         _terminal=self._effect_based_terminal)
+
+    @cached_property
+    def _predicates(self) -> Set[Predicate]:
+        """Helper for initiable and terminal."""
+        return {a.predicate for a in self._operator.preconditions | \
+                self._operator.add_effects | self._operator.delete_effects}
+
+    def _precondition_based_initiable(self, state: State, memory: Dict,
+                                      objects: Sequence[Object],
+                                      params: Array) -> bool:
+        # The memory here is used to store the absolute params, based on
+        # the relative params and the object states.
+        memory["params"] = params  # store for sanity checking in policy
+        changing_objects = [objects[i] for i in self._changing_parameter_idxs]
+        memory["absolute_goal_vec"] = state.vec(changing_objects) + params
+        # Check if initiable based on preconditions.
+        grounded_op = self._operator.ground(tuple(objects))
+        preconditions = grounded_op.preconditions
+        abs_state = utils.abstract(state, self._predicates)
+        return preconditions.issubset(abs_state)
+
+    def _regressor_based_policy(self, state: State, memory: Dict,
+                                objects: Sequence[Object],
+                                params: Array) -> Action:
+        # Compute the updated relative goal.
+        assert np.allclose(params, memory["params"])
+        changing_objects = [objects[i] for i in self._changing_parameter_idxs]
+        relative_goal_vec = memory["absolute_goal_vec"] - state.vec(
+            changing_objects)
+        x = np.hstack(([1.0], state.vec(objects), relative_goal_vec))
+        action_arr = self._regressor.predict(x)
+        assert not np.isnan(action_arr).any()
+        return Action(np.array(action_arr, dtype=np.float32))
+
+    def _effect_based_terminal(self, state: State, memory: Dict,
+                               objects: Sequence[Object],
+                               params: Array) -> bool:
+        del memory, params  # unused
+        # The hope is that we terminate in the effects.
+        grounded_op = self._operator.ground(tuple(objects))
+        abs_state = utils.abstract(state, self._predicates)
+        if grounded_op.add_effects.issubset(abs_state) and \
+            not grounded_op.delete_effects & abs_state:
+            return True
+        # If we fall outside of the space where the preconditions hold, also
```

Comment:
@ashayathalye note: this is a new idea. seems to help with fast failing, at least, which is good for planning.